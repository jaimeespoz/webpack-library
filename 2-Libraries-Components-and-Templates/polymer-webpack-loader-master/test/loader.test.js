/* eslint no-undefined: "off", no-useless-escape: "off" */

const sourceMap = require('source-map');
const loader = require('../src');

const normalisePaths = result => result.replace('src\\\\', 'src/');

async function verifySourceMap(generatedSource, map) {
  const consumer = await new sourceMap.SourceMapConsumer(map);
  consumer.eachMapping((mapping) => {
    if (!mapping.name) {
      return;
    }
    const originalSourceByLine = consumer.sourceContentFor(mapping.source).split('\n');
    const generatedSourceByLine = generatedSource.split('\n');

    expect(generatedSourceByLine[mapping.generatedLine - 1].substr(mapping.generatedColumn, mapping.name.length))
      .toBe(mapping.name);
    expect(originalSourceByLine[mapping.originalLine - 1].substr(mapping.originalColumn, mapping.name.length))
      .toBe(mapping.name);
  });
  consumer.destroy();
}

function addTemplateToPolymerElement(templateValue) {
  return `import {PolymerElement, html} from "@polymer/polymer/polymer-element.js";

class FooElement extends PolymerElement {
  static get is() { return "foo-element"; }
  static get template() { return html\`${templateValue}\`; }
}
customElements.define(FooElement.is, FooElement);
`;
}

describe('loader', () => {
  let opts;

  beforeEach(() => {
    opts = {
      resourcePath: 'src/test.js',
      query: {
        htmlLoader: {
          minimize: false,
          exportAsDefault: true,
          exportAsEs6Default: true,
        },
      },
      sourceMap: false,
    };
  });

  test('can process basic input', (done) => {
    opts.async = () => (err, source, map) => {
      expect(err).toBe(null);
      expect(normalisePaths(source)).toMatchSnapshot();
      expect(map).toBe(undefined);
      done();
    };
    loader.call(opts, 'class Foo {}');
  });

  test('can process without options', (done) => {
    opts.query = null;
    opts.async = () => (err, source, map) => {
      expect(err).toBe(null);
      expect(normalisePaths(source)).toMatchSnapshot();
      expect(map).toBe(undefined);
      done();
    };
    loader.call(opts, 'class Foo {}');
  });

  describe('html-loader', () => {
    test('image sources are replaced with require calls', (done) => {
      opts.async = () => (err, source, map) => {
        expect(err).toBe(null);
        expect(normalisePaths(source)).toMatchSnapshot();
        expect(map).toBe(undefined);
        done();
      };
      loader.call(opts, addTemplateToPolymerElement('<img src="foo.jpg" />'));
    });

    test('html is minimized when option is set', (done) => {
      opts.async = () => (err, source, map) => {
        expect(err).toBe(null);
        expect(normalisePaths(source)).toMatchSnapshot();
        expect(map).toBe(undefined);
        done();
      };
      opts.query.htmlLoader.minimize = true;
      loader.call(opts, addTemplateToPolymerElement(`<div id="foo">
  some text
</div>`));
    });
  });

  describe('styles', () => {
    test('simple styles are left alone', (done) => {
      opts.async = () => (err, source, map) => {
        expect(err).toBe(null);
        expect(normalisePaths(source)).toMatchSnapshot();
        expect(map).toBe(undefined);
        done();
      };
      loader.call(
        opts,
        addTemplateToPolymerElement('<style>* {background-color: transparent;}</style>'));
    });
    test('have url() calls replaced with require statements', (done) => {
      opts.async = () => (err, source, map) => {
        expect(err).toBe(null);
        expect(normalisePaths(source)).toMatchSnapshot();
        expect(map).toBe(undefined);
        done();
      };
      loader.call(opts, addTemplateToPolymerElement('<style>* {background-image: url("foo.jpg");}</style>'));
    });

    test('font url() are properly formatted', (done) => {
      opts.async = () => (err, source, map) => {
        expect(err).toBe(null);
        expect(normalisePaths(source)).toMatchSnapshot();
        expect(map).toBe(undefined);
        done();
      };
      loader.call(opts, addTemplateToPolymerElement(`<style>
        @font-face {
          font-family: 'MyWebFont';
          src: url('webfont.eot'); /* IE9 Compat Modes */
          src: url('webfont.eot?#iefix') format('embedded-opentype'), /* IE6-IE8 */
              url('~webfont.woff2') format('woff2'), /* Super Modern Browsers */
              url('./webfont.woff') format('woff'), /* Pretty Modern Browsers */
              url('webfont.ttf')  format('truetype'), /* Safari, Android, iOS */
              url('webfont.svg#svgFontName') format('svg'); /* Legacy iOS */
        }
        @media print {
          * {background-color: transparent;}
        }
      </style>`));
    });
    test('multiple style tags are processed', (done) => {
      opts.async = () => (err, source, map) => {
        expect(err).toBe(null);
        expect(normalisePaths(source)).toMatchSnapshot();
        expect(map).toBe(undefined);
        done();
      };
      loader.call(
        opts,
        addTemplateToPolymerElement(
          `<style>* {background-image: url("foo.jpg");}</style>
              <style>* {background-image: url("bar.jpg");}</style>`));
    });
    test('missing end tag is skipped', (done) => {
      opts.async = () => (err, source, map) => {
        expect(err).toBe(null);
        expect(normalisePaths(source)).toMatchSnapshot();
        expect(map).toBe(undefined);
        done();
      };
      loader.call(
        opts,
        addTemplateToPolymerElement('<style>* {background-image: url("foo.jpg");}'));
    });
    test('malformed begin tag is skipped', (done) => {
      opts.async = () => (err, source, map) => {
        expect(err).toBe(null);
        expect(normalisePaths(source)).toMatchSnapshot();
        expect(map).toBe(undefined);
        done();
      };
      loader.call(
        opts,
        addTemplateToPolymerElement('<style * {background-image: url("foo.jpg");}'));
    });
    test('@import statements are processed', (done) => {
      opts.async = () => (err, source, map) => {
        expect(err).toBe(null);
        expect(normalisePaths(source)).toMatchSnapshot();
        expect(map).toBe(undefined);
        done();
      };
      loader.call(
        opts,
        addTemplateToPolymerElement(
          '<style>@import url("other.css"); * {background-image: url("foo.jpg");}</style>'));
    });
  });

  describe('full components', () => {
    test('basic element', (done) => {
      opts.query.htmlLoader.minimize = true;
      opts.sourceMap = true;
      opts.async = () => async (err, source, map) => {
        expect(err).toBe(null);
        expect(normalisePaths(source)).toMatchSnapshot();
        expect(map).not.toBe(undefined);
        await verifySourceMap(source, map);
        done();
      };
      loader.call(opts, `
import format from 'date-fns/format';
import {PolymerElement, html} from '@polymer/polymer/polymer-element.js';  

class MyElement extends PolymerElement {
  static get is() { return 'my-element'; }
  static get template() {
    return html\`<h1>Hello, World!</h1>\`;
  }
}

window.customElements.define(MyElement.is, MyElement);
`);
    });

    test('multiple template methods', (done) => {
      opts.query.htmlLoader.minimize = true;
      opts.sourceMap = true;
      opts.async = () => async (err, source, map) => {
        expect(err).toBe(null);
        expect(normalisePaths(source)).toMatchSnapshot();
        expect(map).not.toBe(undefined);
        await verifySourceMap(source, map);
        done();
      };
      loader.call(opts, `
import format from 'date-fns/format';
import {PolymerElement, html} from '@polymer/polymer/polymer-element.js';  

class MyElement extends PolymerElement {
  static get is() { return 'my-element'; }
  static get properties() {
    return {
      today: {
        type: String,
        value: function() {
          return format(new Date(), 'MM/DD/YYYY');
        }
      }
    }
  }
  static get styles() {
    return html\`<style> 
        h1 {
          background-color: pink;
          background-image: url('foo.jpg');
        }
      </style>\`;
  }
  static get template() {
    return html\`
      \${MyElement.styles}
      <h1>
        Hello, World! It's [[today]].
      </h1>\`;
  }
}

window.customElements.define(MyElement.is, MyElement);
`);
    });

    test('html tagged references on same line', (done) => {
      opts.query.htmlLoader.minimize = true;
      opts.sourceMap = true;
      opts.async = () => async (err, source, map) => {
        expect(err).toBe(null);
        expect(normalisePaths(source)).toMatchSnapshot();
        expect(map).not.toBe(undefined);
        await verifySourceMap(source, map);
        done();
      };
      loader.call(opts, `
import format from 'date-fns/format';
import {PolymerElement, html} from '@polymer/polymer/polymer-element.js';  

class MyElement extends PolymerElement {
  static get is() { return 'my-element'; }
  static get styles() {
    return html\`<style>* {background-color: pink;}</style>\`;
  }
  static get template() {
    return html\`<h1>Hello, World!</h1>\` && html\`\${MyElement.styles}\`;
  }
}

window.customElements.define(MyElement.is, MyElement);
`);
    });

    test('source maps account for escaped backtick', (done) => {
      opts.query.htmlLoader.minimize = true;
      opts.sourceMap = true;
      opts.async = () => async (err, source, map) => {
        expect(err).toBe(null);
        expect(normalisePaths(source)).toMatchSnapshot();
        expect(map).not.toBe(undefined);
        await verifySourceMap(source, map);
        done();
      };
      loader.call(opts, `
import format from 'date-fns/format';
import {PolymerElement, html} from '@polymer/polymer/polymer-element.js';  

class MyElement extends PolymerElement {
  static get is() { return 'my-element'; }
  static get styles() {
    return html\`<style>* {background-color: pink;}</style>\`;
  }
  static get template() {
    return html\`foo\\\`s\${MyElement.styles}\`;
  }
}

window.customElements.define(MyElement.is, MyElement);
`);
    });
  });
});
