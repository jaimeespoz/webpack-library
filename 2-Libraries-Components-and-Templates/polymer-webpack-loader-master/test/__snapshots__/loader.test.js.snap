// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loader can process basic input 1`] = `"class Foo {}"`;

exports[`loader can process without options 1`] = `"class Foo {}"`;

exports[`loader full components basic element 1`] = `
"
import format from 'date-fns/format';
import {PolymerElement, html} from '@polymer/polymer/polymer-element.js';  

class MyElement extends PolymerElement {
  static get is() { return 'my-element'; }
  static get template() {
    return html\`<h1>Hello, World!</h1>\`;
  }
}

window.customElements.define(MyElement.is, MyElement);
"
`;

exports[`loader full components html tagged references on same line 1`] = `
"
import format from 'date-fns/format';
import {PolymerElement, html} from '@polymer/polymer/polymer-element.js';  

class MyElement extends PolymerElement {
  static get is() { return 'my-element'; }
  static get styles() {
    return html\`<style>*{background-color:pink}</style>\`;
  }
  static get template() {
    return html\`<h1>Hello, World!</h1>\` && html\`\${MyElement.styles}\`;
  }
}

window.customElements.define(MyElement.is, MyElement);
"
`;

exports[`loader full components multiple template methods 1`] = `
"
import format from 'date-fns/format';
import {PolymerElement, html} from '@polymer/polymer/polymer-element.js';  

class MyElement extends PolymerElement {
  static get is() { return 'my-element'; }
  static get properties() {
    return {
      today: {
        type: String,
        value: function() {
          return format(new Date(), 'MM/DD/YYYY');
        }
      }
    }
  }
  static get styles() {
    return html\`<style>h1{background-color:pink;background-image:url('\${__createTemplateFromString(require(\\"./foo.jpg\\"))}')}</style>\`;
  }
  static get template() {
    return html\`\${MyElement.styles} <h1> Hello, World! It's [[today]]. </h1>\`;
  }
}

window.customElements.define(MyElement.is, MyElement);

function __createTemplateFromString(a) {
  const template = /** @type {!HTMLTemplateElement} */(document.createElement('template'));
  template.content.appendChild(document.createTextNode(a));
  return template;
}
"
`;

exports[`loader full components source maps account for escaped backtick 1`] = `
"
import format from 'date-fns/format';
import {PolymerElement, html} from '@polymer/polymer/polymer-element.js';  

class MyElement extends PolymerElement {
  static get is() { return 'my-element'; }
  static get styles() {
    return html\`<style>*{background-color:pink}</style>\`;
  }
  static get template() {
    return html\`foo\\\\\\\\\\\\\`s\${MyElement.styles}\`;
  }
}

window.customElements.define(MyElement.is, MyElement);
"
`;

exports[`loader html-loader html is minimized when option is set 1`] = `
"import {PolymerElement, html} from \\"@polymer/polymer/polymer-element.js\\";

class FooElement extends PolymerElement {
  static get is() { return \\"foo-element\\"; }
  static get template() { return html\`<div id=\\"foo\\"> some text </div>\`; }
}
customElements.define(FooElement.is, FooElement);
"
`;

exports[`loader html-loader image sources are replaced with require calls 1`] = `
"import {PolymerElement, html} from \\"@polymer/polymer/polymer-element.js\\";

class FooElement extends PolymerElement {
  static get is() { return \\"foo-element\\"; }
  static get template() { return html\`<img src=\\"\${__createTemplateFromString(require('./foo.jpg'))}\\" />\`; }
}
customElements.define(FooElement.is, FooElement);

function __createTemplateFromString(a) {
  const template = /** @type {!HTMLTemplateElement} */(document.createElement('template'));
  template.content.appendChild(document.createTextNode(a));
  return template;
}
"
`;

exports[`loader styles @import statements are processed 1`] = `
"import {PolymerElement, html} from \\"@polymer/polymer/polymer-element.js\\";

class FooElement extends PolymerElement {
  static get is() { return \\"foo-element\\"; }
  static get template() { return html\`<style>@import url('\${__createTemplateFromString(require(\\"./other.css\\"))}'); * {background-image: url('\${__createTemplateFromString(require(\\"./foo.jpg\\"))}');}</style>\`; }
}
customElements.define(FooElement.is, FooElement);

function __createTemplateFromString(a) {
  const template = /** @type {!HTMLTemplateElement} */(document.createElement('template'));
  template.content.appendChild(document.createTextNode(a));
  return template;
}
"
`;

exports[`loader styles font url() are properly formatted 1`] = `
"import {PolymerElement, html} from \\"@polymer/polymer/polymer-element.js\\";

class FooElement extends PolymerElement {
  static get is() { return \\"foo-element\\"; }
  static get template() { return html\`<style>
        @font-face {
          font-family: 'MyWebFont';
          src: url('\${__createTemplateFromString(require(\\"./webfont.eot\\"))}'); /* IE9 Compat Modes */
          src: url('\${__createTemplateFromString(require(\\"./webfont.eot\\"))}?#iefix') format('embedded-opentype'), 
              url('\${__createTemplateFromString(require(\\"webfont.woff2\\"))}') format('woff2'), 
              url('\${__createTemplateFromString(require(\\"./webfont.woff\\"))}') format('woff'), 
              url('\${__createTemplateFromString(require(\\"./webfont.ttf\\"))}')  format('truetype'), 
              url('\${__createTemplateFromString(require(\\"./webfont.svg\\"))}#svgFontName') format('svg'); /* Legacy iOS */
        }
        @media print {
          * {background-color: transparent;}
        }
      </style>\`; }
}
customElements.define(FooElement.is, FooElement);

function __createTemplateFromString(a) {
  const template = /** @type {!HTMLTemplateElement} */(document.createElement('template'));
  template.content.appendChild(document.createTextNode(a));
  return template;
}
"
`;

exports[`loader styles have url() calls replaced with require statements 1`] = `
"import {PolymerElement, html} from \\"@polymer/polymer/polymer-element.js\\";

class FooElement extends PolymerElement {
  static get is() { return \\"foo-element\\"; }
  static get template() { return html\`<style>* {background-image: url('\${__createTemplateFromString(require(\\"./foo.jpg\\"))}');}</style>\`; }
}
customElements.define(FooElement.is, FooElement);

function __createTemplateFromString(a) {
  const template = /** @type {!HTMLTemplateElement} */(document.createElement('template'));
  template.content.appendChild(document.createTextNode(a));
  return template;
}
"
`;

exports[`loader styles malformed begin tag is skipped 1`] = `
"import {PolymerElement, html} from \\"@polymer/polymer/polymer-element.js\\";

class FooElement extends PolymerElement {
  static get is() { return \\"foo-element\\"; }
  static get template() { return html\`<style * {background-image: url(\\"foo.jpg\\");}\`; }
}
customElements.define(FooElement.is, FooElement);
"
`;

exports[`loader styles missing end tag is skipped 1`] = `
"import {PolymerElement, html} from \\"@polymer/polymer/polymer-element.js\\";

class FooElement extends PolymerElement {
  static get is() { return \\"foo-element\\"; }
  static get template() { return html\`<style>* {background-image: url(\\"foo.jpg\\");}\`; }
}
customElements.define(FooElement.is, FooElement);
"
`;

exports[`loader styles multiple style tags are processed 1`] = `
"import {PolymerElement, html} from \\"@polymer/polymer/polymer-element.js\\";

class FooElement extends PolymerElement {
  static get is() { return \\"foo-element\\"; }
  static get template() { return html\`<style>* {background-image: url('\${__createTemplateFromString(require(\\"./foo.jpg\\"))}');}</style>
              <style>* {background-image: url('\${__createTemplateFromString(require(\\"./bar.jpg\\"))}');}</style>\`; }
}
customElements.define(FooElement.is, FooElement);

function __createTemplateFromString(a) {
  const template = /** @type {!HTMLTemplateElement} */(document.createElement('template'));
  template.content.appendChild(document.createTextNode(a));
  return template;
}
"
`;

exports[`loader styles simple styles are left alone 1`] = `
"import {PolymerElement, html} from \\"@polymer/polymer/polymer-element.js\\";

class FooElement extends PolymerElement {
  static get is() { return \\"foo-element\\"; }
  static get template() { return html\`<style>* {background-color: transparent;}</style>\`; }
}
customElements.define(FooElement.is, FooElement);
"
`;
